File: server/app/services/rag.py
from typing import Any, Dict, List
import os

from FlagEmbedding import BGEM3FlagModel
from opensearchpy import OpenSearch


# ---------- OpenSearch & Embedding 초기화 ----------

host = os.getenv("OPENSEARCH_HOST")
port = os.getenv("OPENSEARCH_PORT")
auth = (os.getenv("OPENSEARCH_USER"), os.getenv("OPENSEARCH_PASSWD"))

client = OpenSearch(
    hosts=[{"host": host, "port": port}],
    http_compress=True,
    http_auth=auth,
    use_ssl=True,
    verify_certs=False,
    ssl_assert_hostname=False,
    ssl_show_warn=False,
    timeout=300,
)

INDEX_NAME = os.getenv("OPENSEARCH_INDEX")

# 한 번만 로드해서 재사용
bge_model = BGEM3FlagModel("BAAI/bge-m3", use_fp16=True)


# ---------- 텍스트 청킹 ----------

def chunk_text(pages_text: list[str], chunk_size: int = 500) -> tuple[list[str], list[int]]:
    chunks: list[str] = []
    page_numbers: list[int] = []

    for page_number, text in enumerate(pages_text, start=1):
        words = text.split()
        curr: list[str] = []
      
        for word in words:
            curr.append(word)
            if len(curr) >= chunk_size:
                chunks.append(" ".join(curr))
                page_numbers.append(page_number)
                curr = []

    if curr:
        chunks.append(" ".join(curr)) 
        page_numbers.append(page_number)

    return chunks, page_numbers


# ---------- 임베딩 ----------

def embed_in_batches(chunks: List[str], size: int = 24):
    """chunks 리스트를 받아서 순서대로 dense_vecs 리스트를 반환"""
    all_vecs = []

    for i in range(0, len(chunks), size):
        batch = chunks[i : i + size]

        encoded = bge_model.encode(
            batch,
            batch_size=8,
            max_length=8192,
        )
        # encoded["dense_vecs"] 길이 확인용 로그
        print("encoded batch size:", len(encoded["dense_vecs"]))
        all_vecs.extend(encoded["dense_vecs"])

    return all_vecs

# ---------- 인덱싱 ----------

def index_chunks_for_pdf(
    id: str,
    chunks: List[str],
    dense_vecs: List[list[float]],
    page_numbers: List[int],
) -> None:
    """
    주어진 pdf_id에 대해 (chunk, embedding) 쌍을 OpenSearch에 인덱싱.
    """
    for i, (chunk, vec) in enumerate(zip(chunks, dense_vecs)):
        client.index(
            index=INDEX_NAME,
            id=f"{id}-{i}",
            body={
                "pdf_id": id,
                "chunk_index": i,
                "text": chunk,
                "embedding": vec,
                "page_number": page_numbers[i],
            },
        )

# ---------- KNN 검색 ----------

def search_similar_chunks(
    conversation_id: str,
    query_vec: List[float],
    size: int = 10,
    k: int = 10,
) -> Dict[str, Any]:
    """
    특정 conversation(pdf_id) 내부에서,
    query 임베딩과 가장 유사한 청크들을 OpenSearch로 조회하는 함수.

    Args:
        conversation_id (str): pdf_id (= conversation_id)
        query_vec (List[float]): bge-m3 dense vector (768차원 or 다차원)
        size (int): 반환할 문서 개수
        k (int): knn 이웃 개수

    Returns:
        dict: opensearch search 결과 전체
    """
    body = {
        "size": size,
        "query": {
            "bool": {
                "filter": [
                    {"terms": {"pdf_id": [conversation_id]}}
                ],
                "must": [
                    {
                        "knn": {
                            "embedding": {
                                "vector": query_vec,
                                "k": k
                            }
                        }
                    }
                ]
            }
        }
    }

    result = client.search(
        index=INDEX_NAME,
        body=body,
    )

    return result

File: server/services/rag.ts
import fs from 'fs'
import path from 'path'
import os from 'os'
import { randomUUID } from 'crypto'
import { Client } from '@opensearch-project/opensearch'
import { OpenAIEmbeddings } from '@langchain/openai'
import { PDFLoader } from '@langchain/community/document_loaders/fs/pdf'
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter'

type RagConfig = {
  indexName: string
  opensearchHost: string
  opensearchPort: string
  opensearchUser: string
  opensearchPassword: string
}

function requireEnv(name: string): string {
  const value = process.env[name]
  if (!value) {
    throw new Error(`Missing required environment variable: ${name}`)
  }
  return value
}

function getRagConfig(): RagConfig {
  return {
    indexName: requireEnv('OPENSEARCH_INDEX'),
    opensearchHost: requireEnv('OPENSEARCH_HOST'),
    opensearchPort: requireEnv('OPENSEARCH_PORT'),
    opensearchUser: requireEnv('OPENSEARCH_USER'),
    opensearchPassword: requireEnv('OPENSEARCH_PASSWD'),
  }
}

let opensearchClient: Client | null = null
function getOpenSearchClient(config: RagConfig): Client {
  if (!opensearchClient) {
    opensearchClient = new Client({
      node: `https://${config.opensearchHost}:${config.opensearchPort}`,
      auth: {
        username: config.opensearchUser,
        password: config.opensearchPassword,
      },
      ssl: { rejectUnauthorized: false },
      requestTimeout: 300_000,
    })
  }

  return opensearchClient
}

let embeddings: OpenAIEmbeddings | null = null
function getEmbeddings(): OpenAIEmbeddings {
  if (!embeddings) {
    embeddings = new OpenAIEmbeddings({
      apiKey: process.env.EMBEDDING_API_KEY ?? 'dummy',
      model: '/data/models/baai_bge-m3',
      configuration: {
        baseURL: process.env.EMBEDDING_API_URL,
      },
    })
  }

  return embeddings
}

// ---------- Types ----------

export type SearchHit = {
  text: string
  pageNumber: number
  chunkIndex: number
}

// ---------- Indexing ----------

export async function indexPdf(
  conversationId: string,
  pdfBuffer: Buffer,
): Promise<void> {
  const config = getRagConfig()
  const client = getOpenSearchClient(config)
  const embeddingClient = getEmbeddings()

  const tmpPath = path.join(os.tmpdir(), `${randomUUID()}.pdf`)
  fs.writeFileSync(tmpPath, pdfBuffer)

  try {
    // 1) Parse PDF — one Document per page
    const loader = new PDFLoader(tmpPath, { splitPages: true })
    const docs = await loader.load()

    // 2) Chunk each page's text while preserving page number
    const splitter = new RecursiveCharacterTextSplitter({
      chunkSize: 1000,
      chunkOverlap: 200,
    })
    const chunks = await splitter.splitDocuments(docs)

    if (chunks.length === 0) return

    // 3) Embed all chunks
    const texts = chunks.map((c) => c.pageContent)
    const vectors = await embeddingClient.embedDocuments(texts)

    // 4) Index to OpenSearch
    for (let i = 0; i < chunks.length; i++) {
      const pageNumber =
        (chunks[i].metadata as { loc?: { pageNumber?: number } }).loc
          ?.pageNumber ?? 1
      await client.index({
        index: config.indexName,
        id: `${conversationId}-${i}`,
        body: {
          pdf_id: conversationId,
          chunk_index: i,
          text: chunks[i].pageContent,
          embedding: vectors[i],
          page_number: pageNumber,
        },
      })
    }
  } finally {
    fs.unlinkSync(tmpPath)
  }
}

// ---------- KNN search ----------

export async function searchSimilarChunks(
  conversationId: string,
  query: string,
  size = 10,
  k = 20,
): Promise<SearchHit[]> {
  const config = getRagConfig()
  const client = getOpenSearchClient(config)
  const embeddingClient = getEmbeddings()
  const [queryVec] = await embeddingClient.embedDocuments([query])

  const result = await client.search({
    index: config.indexName,
    body: {
      size,
      query: {
        bool: {
          filter: [{ terms: { pdf_id: [conversationId] } }],
          must: [
            {
              knn: {
                embedding: { vector: queryVec, k },
              },
            },
          ],
        },
      },
    },
  })

  const hits = (result.body.hits?.hits ?? []) as unknown as Array<{
    _source: { text: string; page_number: number; chunk_index: number }
  }>

  return hits.map((h, i) => ({
    text: h._source.text,
    pageNumber: h._source.page_number,
    chunkIndex: h._source.chunk_index ?? i,
  }))
}
